src/hotspot/share/oops/resolvedMethodEntry.hpp

```diff
@@ -0,0 +1,208 @@
+  u2 table_index() const { return _entry_specific._table_index; }
+  u2 constant_pool_index() const { return _cpool_index; }
+  u1 tos_state() const { return _tos_state; }
+  u2 number_of_parameters() const { return _number_of_parameters; }
+  u1 bytecode1() const { return Atomic::load_acquire(&_bytecode1); }
+  u1 bytecode2() const { return Atomic::load_acquire(&_bytecode2); }
+
+  // Flags
+  bool is_vfinal()           const { return (_flags & (1 << is_vfinal_shift))           != 0; }
+  bool is_final()            const { return (_flags & (1 << is_final_shift))            != 0; }
+  bool is_forced_virtual()   const { return (_flags & (1 << is_forced_virtual_shift))   != 0; }
+  bool has_appendix()        const { return (_flags & (1 << has_appendix_shift))        != 0; }
+  bool has_local_signature() const { return (_flags & (1 << has_local_signature_shift)) != 0; }
+
+  bool is_resolved(Bytecodes::Code code) const {
+    switch(code) {
+      case Bytecodes::_invokeinterface:
+      case Bytecodes::_invokehandle:
+      case Bytecodes::_invokespecial:
+      case Bytecodes::_invokestatic:
+        return (bytecode1() == code);
+      case Bytecodes::_invokevirtual:
+        return (bytecode2() == code);
+    default:
+      ShouldNotReachHere();
+      return false;
+    }
+  }
+
+  void adjust_method_entry(Method* new_method) {
+    // this is done during the redefinition safepoint
+    _method = new_method;
+  }
+  bool check_no_old_or_obsolete_entry();
+
+  // Printing
+  void print_on(outputStream* st) const;
+
+  // Setters
+  void set_flags(u1 flags) { _flags |= flags; }
+
+  inline void set_bytecode(u1* code, u1 new_code) {
+  #ifdef ASSERT
+    // Read once.
+    volatile Bytecodes::Code c = (Bytecodes::Code)*code;
+    assert(c == 0 || c == new_code || new_code == 0, "update must be consistent old: %d, new: %d", c, new_code);
+  #endif
+    Atomic::release_store(code, new_code);
+  }
+
+  void set_bytecode1(u1 b1) {
+    set_bytecode(&_bytecode1, b1);
+  }
+
+  void set_bytecode2(u1 b2) {
+    set_bytecode(&_bytecode2, b2);
+  }
+
+  void set_method(Method* m) {
+    Atomic::release_store(&_method, m);
+  }
+
+  void set_klass(InstanceKlass* klass) {
+    _entry_specific._interface_klass = klass;
+  }
+
+  void set_resolved_references_index(u2 ref_index) {
+    _entry_specific._resolved_references_index = ref_index;
+  }
+
+  void set_table_index(u2 table_index) {
+    _entry_specific._table_index = table_index;
+  }
+
+  void set_num_parameters(u2 num_params) {
+    _number_of_parameters = num_params;
+  }
+
+  void fill_in(u1 tos_state, u2 num_params) {
+    _tos_state = tos_state;
+    _number_of_parameters = num_params;
+  }
+
+  void reset_entry();
+
+  // CDS
+  void remove_unshareable_info();
+
+  // Offsets
+  static ByteSize klass_offset()                     { return byte_offset_of(ResolvedMethodEntry, _entry_specific._interface_klass); }
```

Comment:
Perhaps the getters for the entry specific union fields should include an assert to verify that the flags are consistent with the access being performed?